File: solutions/javascript/tournament/1/tournament.js
//
// This is only a SKELETON file for the 'Tournament' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const tournamentTally = (input) => {
    let tableHeader = 'Team                           | MP |  W |  D |  L |  P'
  if (!input || !input.trim())return tableHeader;
  
const matches = input.trim().split('\n');
 const scores = {};
  function initTeam(team) {
  if (!scores[team]) {
    scores[team] = { MP: 0, W: 0, D: 0, L: 0, P: 0 };
  }
}

for (const match of matches) {
  const [team1, team2, result] = match.split(';');
  
    initTeam(team1);
    initTeam(team2);
  scores[team1].MP++;
    scores[team2].MP++;

    if (result === 'win') {
      scores[team1].W++;
      scores[team1].P += 3;
      scores[team2].L++;
    } else if (result === 'loss') {
      scores[team2].W++;
      scores[team2].P += 3;
      scores[team1].L++;
    } else if (result === 'draw') {
      scores[team1].D++;
      scores[team2].D++;
      scores[team1].P++;
      scores[team2].P++;
    }
  }
  const sortedTeams = Object.keys(scores).sort((a, b) => {
    if (scores[b].P !== scores[a].P) return scores[b].P - scores[a].P;
    return a.localeCompare(b);
  });

const tableRows = sortedTeams.map(team => {
  const { MP, W, D, L, P } = scores[team];
  return `${team.padEnd(31)}| ${String(MP).padStart(2)} | ${String(W).padStart(2)} | ${String(D).padStart(2)} | ${String(L).padStart(2)} | ${String(P).padStart(2)}`;
});


  return [tableHeader, ...tableRows].join('\n');
};

File: solutions/python/tournament/1/tournament.py
def tally(rows):
    
    table = {}

    
    for row in rows:

        team1, team2, result = row.split(';')

        
        if team1 not in table:
            table[team1] = {"MP": 0, "W": 0, "D": 0, "L": 0, "P": 0}

        if team2 not in table:
            table[team2] = {"MP": 0, "W": 0, "D": 0, "L": 0, "P": 0}

    
        table[team1]["MP"] += 1
        table[team2]["MP"] += 1

        
        if result == "win":
            table[team1]["W"] += 1
            table[team1]["P"] += 3

            table[team2]["L"] += 1

        elif result == "loss":
            table[team2]["W"] += 1
            table[team2]["P"] += 3

            table[team1]["L"] += 1

        elif result == "draw":
            table[team1]["D"] += 1
            table[team2]["D"] += 1

            table[team1]["P"] += 1
            table[team2]["P"] += 1

    sorted_teams = sorted(
        table.items(),
        key=lambda item: (-item[1]["P"], item[0])
    )

    
    header = "Team                           | MP |  W |  D |  L |  P"
    lines = [header]

    for team, stats in sorted_teams:
        line = (
            f"{team:<31}| "
            f"{stats['MP']:>2} | "
            f"{stats['W']:>2} | "
            f"{stats['D']:>2} | "
            f"{stats['L']:>2} | "
            f"{stats['P']:>2}"
        )
        lines.append(line)

    return lines

